src/win32crypto/keys.c

```diff
@@ -0,0 +1,575 @@
+
+    NTSTATUS err = NCryptSignHash(pk->key,
+                                  flags ? &pad : NULL, hash_bin, hash_bin_len,
+                                  (PBYTE)sig, len, &len, flags);
+    if (err != ERROR_SUCCESS) {
+        LOG_ERROR(ERR, err, "Signature failed");
+        rc = -1;
+    }
+    *siglen = len;
+done:
+
+    if (hash_algo) BCryptCloseAlgorithmProvider(hash_algo, 0);
+    if (hash) BCryptDestroyHash(hash);
+
+    return rc;
+}
+static void free_priv_key(struct tlsuv_private_key_s *key) {
+    struct win32crypto_private_key_s *priv_key = (struct win32crypto_private_key_s *) key;
+    if (priv_key->key) {
+        NCryptFreeObject(priv_key->key);
+        priv_key->key = 0;
+    }
+    if (priv_key->provider) {
+        NCryptFreeObject(priv_key->provider);
+    }
+    tlsuv__free(priv_key);
+}
+
+static struct tlsuv_private_key_s private_key_api = {
+        .to_pem = priv_key_pem,
+        .free = free_priv_key,
+        .get_certificate = NULL,
+        .pubkey = priv_key_pub,
+        .sign = priv_key_sign,
+        .store_certificate = NULL,
+};
+
+static struct win32crypto_private_key_s* new_private_key(NCRYPT_PROV_HANDLE ph, NCRYPT_KEY_HANDLE kh) {
+    struct win32crypto_private_key_s *key = tlsuv__calloc(1, sizeof(*key));
+    key->api = private_key_api;
+    key->provider = ph;
+    key->key = kh;
+    return key;
+}
+
+static wchar_t *providers[] = {
+    MS_PLATFORM_KEY_STORAGE_PROVIDER,
+    MS_KEY_STORAGE_PROVIDER,
+};
+
+static wchar_t *key_algos[] = {
+    NCRYPT_ECDSA_P521_ALGORITHM,
+    NCRYPT_ECDSA_P384_ALGORITHM,
+    NCRYPT_ECDSA_P256_ALGORITHM,
+};
+
+static NCRYPT_PROV_HANDLE get_provider() {
+    NCRYPT_PROV_HANDLE prov = 0;
+    SECURITY_STATUS rc = 0;
+    for (int i = 0; i < sizeof(providers)/sizeof(providers[0]); i++) {
+        rc = NCryptOpenStorageProvider(&prov, providers[i], 0);
+        if (rc == ERROR_SUCCESS) return prov;
+    }
+
+    UM_LOG(ERR, "failed to open supported provider: %s", win32_error(rc));
+    return 0;
+}
+
+static wchar_t* key_name(const char *id) {
+    wchar_t *key_name = tlsuv__calloc(strlen(id) + 1, sizeof(wchar_t));
+    swprintf(key_name, strlen(id) + 1, L"%s", id);
+    return key_name;
+}
+
+int win32crypto_gen_keychain_key(tlsuv_private_key_t *pk, const char *id) {
+    NCRYPT_PROV_HANDLE prov = get_provider();
+
+    if (prov == 0) {
+        UM_LOG(ERR, "failed to open a supported key storage provider");
+        return -1;
+    }
+
+    wchar_t *name = key_name(id);
+    NCRYPT_KEY_HANDLE kh = 0;
+    for (int i = 0; i < sizeof(key_algos)/sizeof(key_algos[0]); i++) {
+        SECURITY_STATUS rc = NCryptCreatePersistedKey(prov, &kh, key_algos[i], name, 0, 0);
+        if (rc != ERROR_SUCCESS) continue;
+
+        rc = NCryptFinalizeKey(kh, 0);
+        if (rc == ERROR_SUCCESS) break;
+
+        NCryptFreeObject(kh);
+        kh = 0;
+    }
+    tlsuv__free(name);
+
+    if (kh) {
+        *pk = (tlsuv_private_key_t)new_private_key(prov, kh);
+        return 0;
```

Comment:
shouldn't NCryptFreeObject be called here?

Reply:
no, the key struct holds the handles